### DISCUSSION TITLE
how to make use all the nodes and works in the cluster

### DISCUSSION THREAD

**Discussion by vihariazure** [milvus-io/milvus#38481] [Q&A and General discussion]
iam running milvus in the data bricks cluster 
i want to make use of all the nodes in the cluster and worker also 
below is my configuration 

![image](https://github.com/user-attachments/assets/a6026661-b128-4310-9290-0058efdb3934)

**Comment by yhmo** [comment]
Sorry we don't have experience to deploy milvus on databricks.

**Reply by vihariazure** [reply]
I have a VM where Milvus is running using Docker.
I am ingesting data from Databricks into Milvus (on the VM).

How can I utilize the Databricks workers to speed up the data ingestion process into Milvus?

**Reply by xiaofan-luan** [reply]
> 

search for Spark Milvus connector and check if that can help

**Comment by xiaofan-luan** [comment]
> iam running milvus in the data bricks cluster i want to make use of all the nodes in the cluster and worker also below is my configuration
> 
> ![image](https://private-user-images.githubusercontent.com/187254384/396004233-a6026661-b128-4310-9290-0058efdb3934.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzMzkyMDEsIm5iZiI6MTczNDMzODkwMSwicGF0aCI6Ii8xODcyNTQzODQvMzk2MDA0MjMzLWE2MDI2NjYxLWIxMjgtNDMxMC05MjkwLTAwNThlZmRiMzkzNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQwODQ4MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OWVkMmFkYjIzY2VhOGNlNTZkNDM3NDlhYTEwMjRiMTAyZjBiZTUzMDZhZTA2MTk3ZTc2NzAzOWJiNzdlN2QzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.YMgJHz1Xeu6ymP_qMAZ5m4tGPTI5A1vAZwGvB6Z5LuQ)

we don't have too too much info about databricks.
can you explain more details about what you want to do?

**Reply by vihariazure** [reply]
we need to speed up data ingestion for that what are the ways 
one way we thought of making utilizing of workers (parallel ingestion ) . how to do that

**Reply by yhmo** [reply]
Basically, the bottleneck of ingestion is not on the client side, it is on the server side.

When you insert some data into milvus by insert():
1. The data is received by proxy node, proxy node splits the data into pieces by shards_num(which is defined in collection schema, default is 1), and sends the pieces(as messages) into Pulsar/Kafka. The insert() returns success to the client once the data is persisted into Pulsar/Kafka.
2. Querynode and datanode subscribe messages asynchronously from Pulsar/Kafka, the pieces of data are consumed by querynode and datanode.
3. Querynode keeps the data in memory for search/query. Datanode persists the data to segments(on s3/minio).

The first bottleneck is the proxy node, it receives data by RPC, which is limited by network bandwidth. You can increase the number of proxies to improve throughput.
The second bottleneck is Pulsar/Kafka
The third bottleneck is querynode/datanode, if datanode consuming speed is much delay to ingestion, milvus will reject any ingestion requests.

So, sometimes even a simple multi-thread client ingestion program could hit the max throughput of a milvus server. The max throughput is affected by network bandwidth, hard disk performance, ability of proxy/querynode/datanode, etc.
You can use spark or some other frameworks to ingest streaming data, but "max ingestion speed" is not a major goal.

"[Import data by files](https://milvus.io/docs/import-data.md)" is a recommended way to import huge amounts of data. In this approach, data file is directly parsed by datanode, Pulsar/Kafka is skipped. With splitting the data into proper size of data files, hopefully the ingestion speed is faster than normal insert requests.

**Reply by xiaofan-luan** [reply]
> we need to speed up data ingestion for that what are the ways one way we thought of making utilizing of workers (parallel ingestion ) . how to do that

import/bulk insert is definitely what you need.
You can import 1B data into milvus in several others
